Add in more generalized support for casting nested types

revans2 opened the pull request:
This fixes #3156 
This does not depend on https://github.com/rapidsai/cudf/pull/8985, but it does help a lot with reducing unneeded data movement.

This adds support for casting structs to structs and maps to maps.  It also generalizes support for casting arrays to arrays so that more types are supported. It also fixed a number of bugs in casting nested types.

I also added in an optimization from the original Spark cast code that checks to see if the types are structurally the same, meaning only the names of the fields in the struct changed, and if they are then it does not need to actually bother with the cast.

revans2 commented:
build

revans2 commented:
build

jlowe commented on shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala:
```diff
@@ -147,13 +147,22 @@ class Spark311Shims extends Spark301Shims {
+        override val arrayChecks: TypeSig =
+          ARRAY.nested(commonCudfTypes + DECIMAL_64 + NULL + ARRAY + BINARY + STRUCT) +
+              psNote(TypeEnum.ARRAY, "The array's child type must also support being cast to " +
+                  "to desired child type")
```
```suggestion
                  "the desired child type")
```

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:
```diff
@@ -1046,41 +1046,50 @@ class CastChecks extends ExprChecks {
+  val arrayChecks: TypeSig = ARRAY.nested(commonCudfTypes + DECIMAL_64 + NULL +
+      ARRAY + BINARY + STRUCT + MAP) +
+      psNote(TypeEnum.ARRAY, "The array's child type must also support being cast to " +
+          "to desired child type")
```
```suggestion
          "the desired child type")
```

jlowe commented on shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala:
```diff
@@ -147,13 +147,22 @@ class Spark311Shims extends Spark301Shims {
-        override val mapChecks: TypeSig = none
+        override val mapChecks: TypeSig =
+          MAP.nested(commonCudfTypes + DECIMAL_64 + NULL + ARRAY + BINARY + STRUCT + MAP) +
+              psNote(TypeEnum.MAP, "the map's kay and value must also support being cast to the " +
```
```suggestion
              psNote(TypeEnum.MAP, "the map's key and value must also support being cast to the " +
```

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:
```diff
@@ -1046,41 +1046,50 @@ class CastChecks extends ExprChecks {
-  val mapChecks: TypeSig = none
+  val mapChecks: TypeSig = MAP.nested(commonCudfTypes + DECIMAL_64 + NULL + ARRAY + BINARY +
+      STRUCT + MAP) +
+      psNote(TypeEnum.MAP, "the map's kay and value must also support being cast to the " +
```
```suggestion
      psNote(TypeEnum.MAP, "the map's key and value must also support being cast to the " +
```

jlowe reviewed (commented):
Minor typos but otherwise lgtm.

revans2 commented:
build

revans2 commented:
build

revans2 commented:
Got some scary test failures. Looking into it more.

revans2 commented:
build

revans2 commented:
There was a small bug in the existing cast code where it was using the member value `dataType` for the to cast type instead of going off of what was passed in. I had already fixed several issues like this, but I didn't know if I had gotten them all. So I moved all of the casting methods to an object so there was no chance of it happening.  The code is a bit more verbose now, because we have to pass around a few boolean flags, but there is no chance of re-introducing the bug without a lot of effort.
